Optimize class_attribute instance accessors

casperisfine opened the pull request:
### Improvement

Here's the benchmark I used: https://gist.github.com/casperisfine/1f37a7228f9133ba30441ff796fe0c1b

```
==== definition ====
 old     11.341k (±24.3%) i/s -     52.415k in   5.032040s
 new      1.708k (±18.9%) i/s -      8.300k in   5.058740s

==== class reader ====
 old     14.143M (± 1.7%) i/s -     70.951M in   5.018039s
 new     14.126M (± 1.4%) i/s -     70.956M in   5.024121s

==== class inherited reader ====
 old     14.196M (± 1.3%) i/s -     71.104M in   5.009397s
 new     13.852M (± 1.4%) i/s -     69.525M in   5.020324s

==== class writer ====
 old    721.934k (±22.4%) i/s -      3.375M in   5.074818s
 new    976.412k (±20.8%) i/s -      4.706M in   5.043937s

==== instance reader ====
 old      5.153M (± 0.9%) i/s -     25.763M in   5.000218s
 new      9.962M (± 1.5%) i/s -     50.054M in   5.025634s

==== instance writer ====
 old      9.411M (± 1.6%) i/s -     47.132M in   5.009605s
 new     21.748M (± 1.5%) i/s -    108.722M in   5.000272s
```

In short, the instance accessors are twice faster with this patch, and the class writer is marginally faster.

The downside is that the definition of the class variable is much slower. So this trade a bit of boot time performance for runtime performance. But it's not like the definition of `class_attribute` is really a hotspot.

### This implementation

Most of the gains are thanks to `class_eval` that allows to replace `instance_variable_defined?(:@foo)` by `defined?(@foo)`, `instance_variable_set(:@foo, value)` by `attr_writer`, and similar changes.

### Discarded implementations

Initially I was actually trying to reclaim some performance on the class accessors. I found an implementation based on `defined?(@foo) ? @foo : superclass.foo`, that was twice faster on the class itself, but then would slowly degrade when calling on child classes.

cc @rafaelfranca @Edouard-chin @etiennebarrie @paracycle @Morriar

byroot committed:
Optimize class_attribute instance accessors

rafaelfranca commented:
@tenderlove do you have any idea how we can improve the class accessor?

casperisfine commented:
Uh weird failures on CI, I didn't expect that since the ActiveSupport test suite was green. I'll take a look.

paracycle commented on activesupport/lib/active_support/core_ext/class/attribute.rb:
```diff
@@ -92,50 +92,63 @@ def class_attribute(
-          end
-        end
-        val
+        method_def <<
```
I think placing the instance predicate method here changes the semantics of this method.

In the code before, the instance predicate was only defined if `instance_reader` was truthy (as well as `instance_predicate`).

In this PR, as long as `instance_predicate` is defined, the instance predicate method is defined, which seems to be a behaviour change.

casperisfine replied:
Nice catch. I fixed it.

casperisfine commented:
CI is green now.

casperisfine commented:
I benched the class accessor against a regular `class << self; attr_reader; end` and it's in the same ballpark:

```
attr_reader     21.296M (± 1.9%) i/s -    106.875M in   
class_attribute reader     14.145M (± 1.7%) i/s -     70.880M in   5.012404s
```

So there's probably no gain to be found on the class reader.

I rebased the PR and fixed the syntax error.

kaspth commented on activesupport/lib/active_support/core_ext/class/attribute.rb:
```diff
@@ -92,50 +92,65 @@ def class_attribute(
+
+    attrs.each do |name|
+      self.public_send("#{name}=", default)
+    end
```
Find the code significantly tougher to read, so here's a potential, albeit only tested in parts, alternative:

```ruby
def class_attribute(*attrs, instance_accessor: true, instance_reader: instance_accessor, instance_writer: instance_accessor, instance_predicate: true, default: nil)
  if name = attrs.find { |name| name.is_a?(Symbol) || name.is_a?(String) }
    raise TypeError, "#{name.inspect} is not a symbol nor a string"
  end

  class_methods, methods = [], []
  attrs.each do |name|
    class_methods << "silence_redefinition_of_method def #{name}; end" # In case the method exists and is not public

    methods << <<~end_of_method if instance_reader
      silence_redefinition_of_method def #{name}
        defined?(@#{name}) ? @#{name} : self.class.#{name}
      end
    end_of_method


    class_methods << <<~end_of_method
      silence_redefinition_of_method def #{name}=(value)
        redefine_method(:#{name}) { value } if singleton_class?
        redefine_singleton_method(:#{name}) { value }
        value
      end
    end_of_method

    methods << <<~end_of_method if instance_writer
      silence_redefinition_of_method(:#{name}=)
      attr_writer :#{name}
    end_of_method


    if instance_predicate
      class_methods << "silence_redefinition_of_method def #{name}?; !!self.#{name}; end"
      methods       << "silence_redefinition_of_method def #{name}?; !!self.#{name}; end"
    end
  end

  location = caller_locations(1, 1).first
  class_eval(["class << self", *class_methods, "end", *methods].join.tr("\n", ";"), location.path, location.lineno)

  attrs.each { |name| public_send("#{name}=", default) }
end
```

Think it's clearer if we group by method kind instead of class then instance. We can also take advantage of `def` returning the method name as a symbol and save a line in most cases (`attr_writer` returns `nil`). Also think it's clearer if the type check is hoisted up top, despite the extra loop. Think there's a significant win with heredoc + tailing ifs.

casperisfine replied:
@kaspth I'm fine with using `.tr("\n", ";")`, but then we should probably apply the same change to `Module#delegate` for consistency's sake. 

cc @rafaelfranca WDYT?

rafaelfranca replied:
I'm fine with that.

casperisfine replied:
Done. With a few adjustments.

I'll see later for Module#delegate.

kaspth commented:
@casperisfine thanks! There were some stylistic things I felt strongly about, so I committed this in https://github.com/rails/rails/commit/49d1b5a98dc50ca59bbfb817f90eb5c2c2645d30

(Accidentally lost the merge commit somehow)
